grype/presenter/sarif/presenter.go

```diff
@@ -0,0 +1,390 @@
+	if inputPath == "." {
+		inputPath = ""
+	}
+	if len(p.Locations) > 0 {
+		location := p.Locations[0]
+		packagePath := location.RealPath
+		if location.VirtualPath != "" {
+			packagePath = location.VirtualPath
+		}
+		if pres.srcMetadata.Scheme == source.DirectoryScheme {
+			packagePath = fmt.Sprintf("%s/%s", inputPath, packagePath)
+		}
+		return packagePath
+	}
+	return inputPath
+}
+
+// locations the locations array is a single "physical" location with potentially multiple logical locations
+func (pres *Presenter) locations(m match.Match) []*sarif.Location {
+	var logicalLocations []*sarif.LogicalLocation
+	physicalLocation := pres.packagePath(m.Package)
+	trimmedPath := strings.TrimPrefix(physicalLocation, "/")
+
+	switch pres.srcMetadata.Scheme {
+	case source.ImageScheme:
+		img := pres.srcMetadata.ImageMetadata.UserInput
+		for _, l := range m.Package.Locations {
+			logicalLocations = append(logicalLocations, &sarif.LogicalLocation{
+				FullyQualifiedName: sp(fmt.Sprintf("%s@%s:/%s", img, l.FileSystemID, trimmedPath)),
+				Name:               sp(l.RealPath),
+			})
+		}
+
+		// this is a hack to get results to show up in GitHub, as it requires relative paths for the location
+		// but we really won't have any information about what Dockerfile on the filesystem was used to build the image
+		// TODO we could add configuration to specify the prefix, a user might want to specify an image name and architecture
+		// in the case of multiple vuln scans, for example
+		physicalLocation = fmt.Sprintf("image/%s", physicalLocation)
+	case source.FileScheme:
+		for _, l := range m.Package.Locations {
+			logicalLocations = append(logicalLocations, &sarif.LogicalLocation{
+				FullyQualifiedName: sp(fmt.Sprintf("%s:/%s", pres.srcMetadata.Path, trimmedPath)),
+				Name:               sp(l.RealPath),
+			})
+		}
+	}
+
+	return []*sarif.Location{
+		{
+			PhysicalLocation: &sarif.PhysicalLocation{
+				ArtifactLocation: &sarif.ArtifactLocation{
+					URI: sp(physicalLocation),
+				},
+				// TODO When grype starts reporting line numbers this will need to get updated
+				Region: &sarif.Region{
+					StartLine:   ip(1),
+					StartColumn: ip(1),
+					EndLine:     ip(1),
+					EndColumn:   ip(1),
+				},
+			},
+			LogicalLocations: logicalLocations,
+		},
+	}
+}
+
+// severityText provides a textual representation of the severity level of the match
+func (pres *Presenter) severityText(m match.Match) string {
+	meta := pres.metadata(m)
+	if meta != nil {
+		severity := vulnerability.ParseSeverity(meta.Severity)
+		switch severity {
+		case vulnerability.CriticalSeverity:
+			return "critical"
+		case vulnerability.HighSeverity:
+			return "high"
+		case vulnerability.MediumSeverity:
+			return "medium"
+		}
+	}
+	return "low"
+}
+
+// cvssScore attempts to get the best CVSS score that our vulnerability data contains
+func (pres *Presenter) cvssScore(v vulnerability.Vulnerability) float64 {
```

Comment:
it would be ideal to have a test explicitly for this (there's a lot packed in here)